x-pack/plugins/ml/common/util/job_utils.ts

```diff
@@ -85,6 +73,29 @@ export function isMappableJob(job: CombinedJob, detectorIndex: number): boolean
   return isMappable;
 }
 
+export function hasValidComposite(buckets: { [key: string]: any }) {
+  if (
+    isPopulatedObject(buckets, ['composite']) &&
+    isPopulatedObject(buckets.composite, ['sources'])
+  ) {
+    const sources = buckets.composite.sources as Array<Record<string, object>>;
+    return sources
+      .map((source) => {
+        const sourceName = getFirstKeyInObject(source);
+        if (sourceName !== undefined && isPopulatedObject(source[sourceName])) {
+          const sourceTypes = Object.keys(source[sourceName]);
+          return (
+            sourceTypes.length === 1 &&
+            (sourceTypes[0] === 'date_histogram' || sourceTypes[0] === 'terms')
+          );
+        }
+        return false;
+      })
+      .every((isValidCompositeSource: boolean) => isValidCompositeSource === true);
+  }
+  return true;
```

Comment:
just to clarify the intention of this function, does the missing composite suppose to be valid?

Reply:
Yes I think it makes sense here if the datafeed doesn't have a composite, it should be valid. Only when it does have "composite" defined that it should satisfy the criteria.